### ISSUE TITLE
Wrong Classification with custom model

### ISSUE THREAD

**Issue by diogogonnelli** [zezo357/pytorch_lite#43]
Hello everyone. 

I'm testing the implementation with a model trained by myself, and I'm facing some issues during the classification process. I had to modify the code to export the YOLO classification model, and I did it like this:

```
import torch
import os
from torch import *
from torch.utils.mobile_optimizer import optimize_for_mobile

dev = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")

model = torch.hub.load('main_folder', 'custom', 'yolov5s-cls.pt', source='local')
model.cpu()

example = torch.zeros((1, 3, 224, 224))
traced_script_module = torch.jit.trace(model, example)
optimized_traced_model = optimize_for_mobile(traced_script_module)
optimized_traced_model._save_for_lite_interpreter("model.pt")
```

It seems like the export is working, but when I use it in Flutter, it always gets classified with the wrong class.\

In my case, my model was trained with only 3 classes, and I updated your code to accommodate this number of classes.

I also calculated the mean and standard deviation based on my dataset.

In addition to changing the number of classes in the code, do I need to make any other modifications in the code to use a custom model with a different number of classes?

**Comment by zezo357** [comment]
Can you share your code, Also did you update the labels.txt?

And for yolo no need to change the model the mean and std if you're using the default values, so can you test with the default?

**Comment by diogogonnelli** [comment]
Certainly, I will provide the codes, model, and classes along with the image I am testing.
[model.zip](https://github.com/zezo357/pytorch_lite/files/12665901/model.zip)

Yes, I tested it with the mean and standard deviation values as well.

For this attached image, the return class should be 'andorinha', but regardless of which image I use, it always returns the class 'objeto'.

Here are some parts of my code.



https://github.com/zezo357/pytorch_lite/assets/34580126/769e4529-11a3-493c-bbe7-30b4384f5e5f


![MicrosoftTeams-image (38)](https://github.com/zezo357/pytorch_lite/assets/34580126/42778c6e-64bc-4a95-b69e-8b0cedd77501)

![MicrosoftTeams-image (37)](https://github.com/zezo357/pytorch_lite/assets/34580126/04273c97-104e-4ea7-aff7-3fa70018e70f)

**Comment by zezo357** [comment]
Okay I see that all of your codes are correct, but I can't figure out why the results are not correct,



Anyway I am working on a new release for iOS support and changing how the package reads and gets results so can you wait for new release? 


When I do I can further test why the classification is not correct, and improve it if I can.

**Comment by diogogonnelli** [comment]
Of course, I will wait and conduct new tests with the new build. Thank you for your attention!

**Comment by diogogonnelli** [comment]
Hello @zezo357, were you able to reproduce my code and test it? Do you need any additional information? If you can help with anything, please let me know, as I ran some tests and still have the same result. Thanks!

**Comment by zezo357** [comment]
I was not free lately, thank you for reminding me to test again sry for the wait 



I did try some stuff and changed variables without any useful results, so I will try some different normalization and see what works and let you know, still no promises that I will be able to make it work

**Comment by diogogonnelli** [comment]
Meanwhile, I will continue running some tests. If there is progress, I will let you know.

**Comment by diogogonnelli** [comment]
> > Meanwhile, I will continue running some tests. If there is progress, I will let you know.
> > To improve reproduction, here are the codes used, original model, and converted model, along with the output tensors.
> > https://drive.google.com/file/d/149grCV5f6vQkXw9j4xoDufcQN6LW3z1I/view?usp=share_link
> 
> i have sent a request to access, hopefully i can figure something this weekend

Access granted. Any news let me know. Thanks

**Comment by zezo357** [comment]
can you guys try on this branch?

it maybe be the fix for the problem not sure, but lets see 
```
  pytorch_lite:
    git: 
      url: https://github.com/zezo357/pytorch_lite
      ref: fix-classification-bug
```

**Comment by zezo357** [comment]
Also sorry for asking but, can you provide a GitHub repo i can test on, instead of asking you to test? 

Just to be more efficient while testing and having the same configuration

**Comment by ebdjesus** [comment]
@zezo357 hello my friend!

How you doing?

I have shared with you the code we used to test pytorch_lite

https://github.com/ebdjesus/ri_novo/

**Comment by zezo357** [comment]
> @zezo357 hello my friend!
> 
> How you doing?
> 
> I have shared with you the code we used to test pytorch_lite
> 
> https://github.com/ebdjesus/ri_novo/

thank you will test and let you know if new fix is enough

**Comment by ebdjesus** [comment]
> > @zezo357 hello my friend!
> > How you doing?
> > I have shared with you the code we used to test pytorch_lite
> > https://github.com/ebdjesus/ri_novo/
> 
> thank you will test and let you know if new fix is enough

I'm sending the images in case I need to test with this model we generated

![produtos_andorinha](https://github.com/zezo357/pytorch_lite/assets/51535131/eb7b3a52-1bee-492e-89ae-e012fff42783)

![123](https://github.com/zezo357/pytorch_lite/assets/51535131/3e79a146-94fb-4bdd-9077-b5bfa4a48d85)

**Comment by zezo357** [comment]
@ebdjesus i cant find a way to test the model? the code is not using it? 

the camera icon prints the path, but doesnt run the classification function 
<img width="351" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/d152240f-6329-4e8e-b1dc-fcef1e5cc435">

<img width="901" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/7db95847-cfc7-41b2-9ce3-4d44adb3f0c1">

**Comment by zezo357** [comment]
can you let me know how to run the model in your app ?

**Comment by ebdjesus** [comment]
After capturing the image, you need to click the + button to run the classification.

<img width="451" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/51535131/5e0900f9-8b8f-4588-b192-b5ec3d54365f">

**Comment by zezo357** [comment]
i just tested on the package example (just replacing model and labels files )


<img width="1171" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/01670992-8212-447a-a9e3-de8260c9c56a">
<img width="1141" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/db85c930-1919-4dcd-8082-947816b0baed">
<img width="1156" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/3b64431f-04cc-4b49-a062-c721d8feb086">


all of the results are correct? where is the problem?

**Comment by zezo357** [comment]
my function 
```dart 
    final XFile? image = await _picker.pickImage(source: ImageSource.gallery);
    //get prediction
    //labels are 1000 random english words for show purposes
    print(image!.path);
    Stopwatch stopwatch = Stopwatch()..start();

    String textToShow = await _imageModel!
        .getImagePrediction(await File(image.path).readAsBytes());
    textToShow = "${textToShow ?? ""}, ${inferenceTimeAsString(stopwatch)}";

```

**Comment by ebdjesus** [comment]
this image is gallo label its correct

Now try with this image that is andorinha label

![123](https://github.com/zezo357/pytorch_lite/assets/51535131/912deff0-9183-4067-b819-6f293b2ced9e)

**Comment by zezo357** [comment]
> this image is gallo label its correct
> 
> Now try with this image that is andorinha label
> 
> ![123](https://user-images.githubusercontent.com/51535131/276985894-912deff0-9183-4067-b819-6f293b2ced9e.png)

also i tested this image and it works too ?? ?

<img width="380" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/576a28d7-0249-4ef0-b916-a1bd2995a8b3">

**Comment by zezo357** [comment]
in your code you have 
<img width="535" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/25157308/db27c888-5d4d-4972-b68c-c2ab01785e3e">

i am thinking this is the problem, since using the default works correctly

**Comment by ebdjesus** [comment]
@zezo357 
all your tests are correct

 I downloaded your code and included the branch you provided and the models and label I trained, but I was unsuccessful.

![image](https://github.com/zezo357/pytorch_lite/assets/51535131/299a3096-baf5-4270-b9fd-46fa259336be)

<img width="1149" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/51535131/21432c0c-a356-4f4b-bac9-6cc2c31513b1">


<img width="445" alt="image" src="https://github.com/zezo357/pytorch_lite/assets/51535131/9bd70521-2edd-4c24-a41e-8ef5169b3857">

**Comment by zezo357** [comment]
Did you close app and doing a pub get and rerun it after changing the models?

Please set it as a different name to make sure it's replaced correctly

**Comment by zezo357** [comment]
https://github.com/zezo357/pytorch_lite/tree/classifer-test this is the code i tested on and showed the results to you

**Comment by ebdjesus** [comment]
I deleted the .pub-cache and tried to clear everything, changed the name as you said and nothing.
I was creating a new project to import your classes.

**Comment by zezo357** [comment]
The model I used is the one in your repo, and it works correctly, 


And I did give you the branch I tested on, this is weird


Please can you try on the branch I provided?

**Comment by ebdjesus** [comment]
@zezo357 

I have now tested it directly in my project and it is working very well.
The problem was that @diogogonnelli  had given me the old classification file, I'm going to kick his ass here ;D
Once again thank you very much for your support

**Comment by zezo357** [comment]
Glad to help and good luck , also don't be too harsh on him we all mess up sometimes. All we can do is learn from mistakes.

**Comment by ebdjesus** [comment]
I'm just kidding, he's largely responsible for us having this functionality today and you for making it possible to run this in the app.

**Comment by zezo357** [comment]
> I'm just kidding, he's largely responsible for us having this functionality today and you for making it possible to run this in the app.

Hopefully all goes well, and don't hesitate to ask for assistance if needed
